clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```
@@ -0,0 +1,591 @@
+                        // if this topic partition of this consumer no longer exists remove it from currentAssignment of the consumer
+                        partitionIter.remove();
+                        currentPartitionConsumer.remove(partition);
+                    } else
+                        // otherwise, remove the topic partition from those that need to be assigned (because it is already assigned
+                        // and we would want to preserve that assignment as much as possible)
+                        unassignedPartitions.remove(partition);
+                }
+            }
+        }
+        // at this point we have preserved all valid topic partition to consumer assignments and removed
+        // all invalid topic partitions and invalid consumers. Now we need to assign unassignedPartitions
+        // to consumers so that the topic partition assignments are as balanced as possible.
+
+        // an ascending sorted set of consumers based on how many topic partitions are already assigned to them
+        TreeSet<String> sortedCurrentSubscriptions = new TreeSet<>(new SubscriptionComparator(currentAssignment));
+        sortedCurrentSubscriptions.addAll(currentAssignment.keySet());
+
+        balance(sortedPartitions, unassignedPartitions, sortedCurrentSubscriptions, consumer2AllPotentialPartitions,
+                partition2AllPotentialConsumers, oldAssignment, currentPartitionConsumer);
+        return currentAssignment;
+    }
+
+
+    @Override
+    protected ByteBuffer getUserData() {
+        try {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(2048);
+            ObjectOutputStream oos = new ObjectOutputStream(bos);
+            oos.writeObject(currentAssignment);
+            return ByteBuffer.wrap(bos.toByteArray());
+        } catch (Exception ioex) {
+            log.error("Failed to serialize currentAssignment", ioex);
+            return super.getUserData();
```

Might be better to just propagate the exception and fail. We wouldn't expect to actually hit this in practice.

---

Wouldn't this require modifying the method signature in the super class too? I assume you meant adding a something like `... throws IOException`.